src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
     }
   }
 
-
   public Map<TopicPartition, Long> getPreviousOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
+  public void writeRecord(final TopicPartition topicPart, final SinkRecord record)
+      throws CopycatException, InterruptedException {
+    lock.lock();
     try {
+      recover(topicPart);
+
       RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
       writer.write(System.currentTimeMillis(), record);
+
       if (!offsets.containsKey(topicPart)) {
         offsets.put(topicPart, record.getOffset() - 1);
       }
       updateRecordCounter(topicPart);
+
+      if (!tasks.containsKey(topicPart)) {
+        ScheduledFuture task = timedRotateThread.schedule(new Runnable() {
+          @Override
+          public void run() {
+            lock.lock();
+            try {
+              rotate(topicPart);
+            } catch (IOException | InterruptedException e) {
+              throw new CopycatRuntimeException(e);
+            } finally {
+              lock.unlock();
+            }
+          }
+        }, rotateInterval, TimeUnit.SECONDS);
+        tasks.put(topicPart, task);
+      }
       if (shouldRotate(topicPart)) {
         rotate(topicPart);
       }
     } catch (IOException e) {
       throw new CopycatRuntimeException(e);
+    } finally {
+      lock.unlock();
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
+  public void close() throws IOException, CopycatException, InterruptedException {
+    timedRotateThread.shutdown();
+
+    for (TopicPartition topicPart : writers.keySet()) {
       rotate(topicPart);
     }
     writers.clear();
+
+    for (WAL wal : wals.values()) {
+      wal.close();
+    }
+    wals.clear();
+
+    tempFileNames.clear();
+    recordCounters.clear();
+    offsets.clear();
+    tasks.clear();
     if (fs != null) {
       fs.close();
     }
   }
 
-  private Map<TopicPartition, Long> readOffsets() throws IOException {
-    Map<TopicPartition, Long> offsets = new HashMap<TopicPartition, Long>();
+  private void recover(TopicPartition topicPart) throws IOException {
+    if (!recovered.contains(topicPart)) {
+      WAL wal = new WAL(conf, url, topicsDir, topicPart);
+      wal.apply();
+      wals.put(topicPart, wal);
+      recovered.add(topicPart);
+      readOffsets(topicPart);
+    }
+  }
+
+  private void readOffsets() throws IOException {
     Path path = new Path(url + "/" + topicsDir);
-    PathFilter filter = new CommittedFileFilter();
     FileStatus[] topics = fs.listStatus(path);
-    for (FileStatus topic: topics) {
+    for (FileStatus topic : topics) {
       if (topic.isDirectory()) {
         String topicName = topic.getPath().getName();
-        FileStatus[] partitions = fs.listStatus(topic.getPath(), filter);
+        FileStatus[] partitions = fs.listStatus(topic.getPath());
```

Comment:
First, it may be worth retrying -- if it's a really short transient issue, you might be able to recover quickly. Besides that, yes, I think we want to do what you said but probably throw a CopycatException also since the callers of this method *require* that it finish loading offsets. When we can indicate that messages should be redelivered in a subsequent put() call, then we can use that exception type and then we will retry automatically with the next iteration (which could also sleep or something if it needs to back off).

I'm realizing that exposing pause/resume methods of the consumer will also be important -- we don't want one topic partition getting stuck for some reason to affect them all.